### DISCUSSION TITLE
Feedback thread for `set_pipeline` step with `teams:`

### DISCUSSION THREAD

**Discussion by vito** [concourse/concourse#5731] [Show and tell]
This thread is to gather feedback for the use of the `teams:` field with the `set_pipeline` step.

(I'll circle back here and add some example questions/prompts, this is just a placeholder for now so we can link to it from the PR that adds it.)

**Comment by BalrajMomi** [comment]
I found this feature extremely helpful in completely automating the pipeline updates. I personally prefer this solution over previous solution which required us to supply username and password.

Now comes the sad part about using it at scale. I plan to reuse same template for multiple projects by externalizing all of my pipeline variables.

**set_pipeline**: requires name of the pipeline to be hardcoded. When I tried giving a variable ((pipeline)), It created a new pipeline with label ((pipeline)). Given this issue, I am having to fall back to legacy option.

**Reply by vito** [reply]
@BalrajMomi You may want to follow https://github.com/concourse/concourse/issues/5808 which is intended to replace the "super long and ugly generated pipeline names" pattern. 🙂 @mouellet is working on it in this PR: https://github.com/concourse/concourse/pull/5896

**Comment by Krenair** [comment]
This is pretty helpful to us, but it'd be more clear what step is what if the output contained the name of the team the pipeline is being set in:
![image](https://user-images.githubusercontent.com/826662/90162700-f263ad00-dd8c-11ea-8bec-5c00bc4f52f4.png)
(each of those is a different team)

**Reply by vito** [reply]
Good point. Would you mind providing some context as to what these pipelines are doing? My working theory is that `set_pipeline` with `team` shouldn't really exist forever, and that we need to provide better ways for teams to self-automate in some sort of hierarchy, so knowing how you're using this feature will help me know whether your use case can be satisfied by those future plans.

For example, this comment shows how one team may be able to propagate pipelines to other teams: https://github.com/concourse/concourse/pull/5729#discussion_r438950978

(Thanks!)

**Reply by Krenair** [reply]
Basically this pipeline is setting an `info` pipeline (same name and same pipeline YAML file) in each different team - this `info` pipeline knows how to extract certain per-team information about our setup such as cloud provider role names, IP addressing, and various services we provision per-Concourse-team like object storage and docker image repositories. That information is only available to the worker nodes for that team.

**Reply by juanluis-garrote** [reply]
In our case DevOps team (which is participated by every team) owns the main "set-all-pipelines" pipeline. For that it needs to be able to set pipelines for all teams. From there we luckily have the same needs all over the place, and we don't actually need the teams to be able to set pipelines other than the standard ones provided so far. At the end of the day they are part of DevOps team and their needs are considered from day 1.

I get your point of view on dividing domains and letting teams set their things but I can see there is no way to cascade the creation of all our pipelines from a single command as we have now unless we are able to set pipelines for all teams.

**Comment by ghost** [comment]
We use the teams field to allow us to promote versions of pipelines to different environments i.e. promote a version to sandbox then to preprod then prod. 
Every environment has its own "team" to facilitate a clear separation of pipelines. The pipeline that sets the others is currently in the control-plane team and not the main team. To be honest I'm not sure how else we would do it if the teams field disappeared. 
If we could promote the control-plane team to an admin team I think that could work but currently there is no way to do that AFAIK.

**Comment by jamie-pate** [comment]
My use case is configuration.

I'm provisioning concourse to replace jenkins, and concourse has no 'state' which means that I want to use `set_pipeline` in a meta-pipeline to provision all the various pipelines from a central repository.

I've got the pipelines for all teams being provisioned from one 'pipelines' pipeline. I'm only using the teams to segregate secret stores at this point. Once I'm onboarding more users it will be important that each team's secrets are isolated from users who are not in that team.

I suppose I would have to create multiple meta-pipelines if this setting weren't available, and a script that initialized the `team-pipelines` meta-pipeline for each team? I could see where splitting each team's pipelines into a separate 'set-pipelines' pipeline for each team could make sense, but at the current scale I'm working with this would be more effort.

I'm also not clear as to why this would go away, or be considered a bad practice?

**Comment by akath19** [comment]
Our use case is gitops for our pipelines, we have several teams (one per development team per environment) and we use the ```team``` field to have a single pipeline that configures everything, most of our pipelines are very similar (our teams use only 1 programming language) and a change in the general structure is usually propagated to 200+ pipelines at the same time.

Would there be a replacement for the ```team``` field that would allow us to achieve this?

**Comment by BillBennett** [comment]
AFAIK teams are the only way (other than instance vars) to group pipelines in the UI and so ironically using teams seems to be the only way to reduce the visual clutter that instanced pipelines and RFC32 (projects) creates.

In my case I have one pipeline that sets up an instanced pipeline (and yes we've also been experimenting permissions boundary between the two as others already said). I don't think these pipelines can be merged since that would required instanced _jobs_ inside a non-instanced pipeline.

So if I put them in different teams i can hide the mechanics (concourse internals) of how the instanced pipeline came to be so the team can focus on the actual instances.

**Comment by Arkessler** [comment]
The `team` flag on `set_pipeline` is fulfilling a critical role in our concourse gitops flow. It lets us have a locked down pipeline in the main team that deploys all other pipelines for the other teams. Would love to see it moved to a supported feature so that we can rest a bit easier relying on it :)

**Comment by taylorsilva** [comment]
`team` field is not going anywhere. I think we should specify the `team` in the step name as per https://github.com/orgs/concourse/discussions/5731#discussioncomment-50736
I'll open up an issue for this. I'm closing and locking this issue.
